File: LEETCODE/LEETCODE-CPP/113.cpp
//113. Path Sum II
//https://leetcode.com/problems/path-sum-ii/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> res;
    
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<int> temp;
        
        calc(root,targetSum,temp);
        return res;
    }
    
    void calc(TreeNode* root, int ts, vector<int> temp){
        if(!root)
            return;
        
        temp.push_back(root->val);
        ts-=root->val;
        
        if(!root->left && !root->righ){
            if(ts==0)
                res.push_back(temp);
        }
        else{
            calc(root->left, ts, temp);
            calc(root->right, ts, temp);
        }
         
        temp.pop_back();
    }
};

File: LEETCODE/LEETCODE-PYTHON/113.py
# 113. Path Sum II
# https://leetcode.com/problems/path-sum-ii/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#Method-1 using DFS
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        res=[]
        self.dfs(root, targetSum, res, [])

        return res

    def dfs(self, root, targetSum, res, temp):
        if not root:
            return

        temp.append(root.val)

        if not root.left and not root.right and targetSum==sum(temp):
            res.append(temp[:])

        self.dfs(root.left, targetSum, res, temp)
        self.dfs(root.right, targetSum, res, temp)
        temp.pop()